design/cluster-api-provider-metal3/node_reuse.md

```diff
@@ -0,0 +1,200 @@
+flow of host selection.
+
+## Design Details
+
+We propose modifying the Metal3MachineTemplate CRD to support enabling/disabling
+node reuse feature. Add `nodeReuse` field under the spec of the Metal3MachineTemplate,
+that stores boolean type of a value.
+
+When set to True, CAPM3 machine controller tries to reuse the same pool of hosts.
+
+If no value is provided by the user, default value False is set and the current
+flow is preserved.
+
+E.g. Metal3MachineTemplate CR
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3MachineTemplate
+spec:
+  nodeReuse: True
+  template:
+    spec:
+      image:
+        ...
+```
+
+During upgrade, to avoid our hosts from being selected for other MD/KCP pools,
+`infrastructure.cluster.x-k8s.io/node-reuse` label will be set. The label
+stores the name of the KCP/MD that hosts belong to.
+
+E.g. BareMetalHost CR
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+    labels:
+      infrastructure.cluster.x-k8s.io/node-reuse: md-pool1 # name of MD
+```
+
+### Implementation Details/Notes/Constraints
+
+#### Node Reuse Approach
+
+We should perform two steps to re-use the same pool of hosts.
+
+1. During host deprovisioning, set the `infrastructure.cluster.x-k8s.io/node-reuse`
+  label;
+2. During next provisioning, filter out hosts with the label.
+
+The actual implementation will be done within the CAPM3 Machine controller.
+
+Step1 can be done as follows:
+
+- CAPM3 controller sets the `infrastructure.cluster.x-k8s.io/node-reuse: md-pool1`
+  label on the hosts belonging to the same MD/KCP.
+
+Step2 can be done as follows:
+
+- CAPM3 controller filters host labelled with
+  `infrastructure.cluster.x-k8s.io/node-reuse: md-pool1` label.
+
+  Next:
+  - If host is found in `Ready` state:
+    - Pick that host for newly created M3M;
+    - Once it is picked up, remove the whole label from the host.
+  - If host is found in `Deprovisioning` state:
+    - Requeue until that host becomes `Ready`;
+  - If no host is found, while it should be (i.e for some reason host
+    is not in the cluster anymore):
+    - Fall back to the current flow, which selects host randomly.
+
+**What if there are two parallel MD/KCP experiencing upgrade?**
+
+  Two MachineDeployments(md-pool1 and md-pool2) with nodeReuse set to True & False
+  respectively start upgrade at the same time. For md-pool1, CAPM3 Machine
+  controller tries to find hosts with
+  `infrastructure.cluster.x-k8s.io/node-reuse: md-pool1` label, while for md-pool2,
+  it selects any random host, but not those with
+  `infrastructure.cluster.x-k8s.io/node-reuse: md-pool1` label that belong to md-pool1.
+
+**What if we found N-1 hosts with label, while we expect N hosts?**
+  Then we select any other hosts, even though it has no label we expect.
+
+The following diagram shows the example flow of node reuse feature in details:
+
+![node_reuse diagram](../images/node_reuse.png)
```

Comment:
I've added some additional specific comments for the diagram directly in the source file. In general I think it could really help, but the current version needs to be worked out a little bit more to remove some ambiguities and simplify it.

Reply:
Hi @andfasano 
Yes sorry for that. We also felt the same. And we got the review just before pushing the updated versions :).
Please take a look once more, and let us know if they need to be updated. Thanks